Enable step_param for Adam functional optimizer

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62611

Enables optimizer overlap with backwards in DDP for Adam. Additional optimizers, especially Adagrad will be done in follow up diffs.

1. Implement `step_param` method based on `step` in _FunctionalAdam (perf permitting we can later dedupe `step` to call `step_param`
2. Modify tests to test all current functional optimizers.

Differential Revision: [D29891783](https://our.internmc.facebook.com/intern/diff/D29891783/)

rohan-varma committed:
Enable step_param for Adam functional optimizer

Enables optimizer overlap with backwards in DDP for Adam. Additional optimizers, especially Adagrad will be done in follow up diffs.

1. Implement `step_param` method based on `step` in _FunctionalAdam (perf permitting we can later dedupe `step` to call `step_param`
2. Modify tests to test all current functional optimizers.

Differential Revision: [D29891783](https://our.internmc.facebook.com/intern/diff/D29891783/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable step_param for Adam functional optimizer"


Enables optimizer overlap with backwards in DDP for Adam. Additional optimizers, especially Adagrad will be done in follow up diffs.

1. Implement `step_param` method based on `step` in _FunctionalAdam (perf permitting we can later dedupe `step` to call `step_param`
2. Modify tests to test all current functional optimizers.

Differential Revision: [D29891783](https://our.internmc.facebook.com/intern/diff/D29891783/)

[ghstack-poisoned]

SciPioneer commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -52,6 +52,11 @@
 if not IS_WINDOWS:
     from torch.distributed.optim.functional_sgd import _FunctionalSGD
+    from torch.distributed.optim.functional_adam import _FunctionalAdam
+    _SUPPORTED_OPTIM_MAPPING = {
```
Do you think it will be better to move this mapping to `torch.distributed.optim.__init__`? Depending on whether it will be used by users.

rohan-varma replied:
Yes, it is duplicated in a few places now. Will do it in a follow up PR if you don't mind.

SciPioneer commented on torch/distributed/optim/functional_adam.py:
```diff
@@ -45,13 +46,62 @@ def __init__(
+            state['exp_avg'] = torch.zeros_like(param, memory_format=torch.preserve_format)
+            state['exp_avg_sq'] = torch.zeros_like(param, memory_format=torch.preserve_format)
+            if self.amsgrad:
+                state['max_exp_avg_sq'] = torch.zeros_like(param, memory_format=torch.preserve_format)
```
After this line, will `self.state[param]` be updated as well? If not, will `state` be overwritten as an empty dict in L80?

rohan-varma replied:
Yeah, we can think of `self.state[param]` as a dict of dicts. For example in the following program:
```
d = {}
if 1 not in d:
    d[1] = {}
    state = d[1]
    state[1] = 1

state = d[1]
print(state)
```
gives
` >>> {1: 1}`
as expected

The same thing is also done in `step()` function in this file (we have an issue to dedupe the implementations as a bootcamp task).

SciPioneer replied:
Thanks for the explanation!

SciPioneer commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -1704,12 +1713,57 @@ def test_fp16_compress_wrapper_nccl(self):
     @requires_nccl()
     @skip_if_lt_x_gpu(2)
     def test_hook_then_optimizer_nccl(self):
```
Nit: rename it as `test_hook_then_sgd_nccl`?

SciPioneer commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -1704,12 +1713,57 @@ def test_fp16_compress_wrapper_nccl(self):
 
     @requires_nccl()
     @skip_if_lt_x_gpu(2)
     def test_hook_then_optimizer_nccl_grad_as_bucket_view(self):
```
Nit: rename it as `test_hook_then_sgd_nccl_grad_as_bucket_view`?

rohan-varma committed:
Update on "Enable step_param for Adam functional optimizer"


Enables optimizer overlap with backwards in DDP for Adam. Additional optimizers, especially Adagrad will be done in follow up diffs.

1. Implement `step_param` method based on `step` in _FunctionalAdam (perf permitting we can later dedupe `step` to call `step_param`
2. Modify tests to test all current functional optimizers.

Differential Revision: [D29891783](https://our.internmc.facebook.com/intern/diff/D29891783/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable step_param for Adam functional optimizer"


Enables optimizer overlap with backwards in DDP for Adam. Additional optimizers, especially Adagrad will be done in follow up diffs.

1. Implement `step_param` method based on `step` in _FunctionalAdam (perf permitting we can later dedupe `step` to call `step_param`
2. Modify tests to test all current functional optimizers.

Differential Revision: [D29891783](https://our.internmc.facebook.com/intern/diff/D29891783/)

[ghstack-poisoned]
